File: typespec/hub/posts.ts
import { Handle } from "../common/common";
import { EmployerPost } from "../common/posts";
import { VTagID } from "../common/vtags";

export class AddFTPostRequest {
  content: string = "";
  tag_ids: VTagID[] = [];

  IsValid(): boolean {
    return (
      this.content.length > 0 &&
      this.content.length <= 255 &&
      this.tag_ids.length <= 3
    );
  }
}

export class AddPostRequest {
  content: string = "";
  tag_ids: VTagID[] = [];

  IsValid(): boolean {
    return (
      this.content.length > 0 &&
      this.content.length <= 4096 &&
      this.tag_ids.length <= 3
    );
  }
}

export interface AddPostResponse {
  post_id: string;
}

export interface Post {
  id: string;
  content: string;
  tags: string[];
  author_name: string;
  author_handle: Handle;
  created_at: string;
  upvotes_count: number;
  downvotes_count: number;
  score: number;
  me_upvoted: boolean;
  me_downvoted: boolean;
  can_upvote: boolean;
  can_downvote: boolean;
  am_i_author: boolean;
  can_comment: boolean;
  comments_count: number;
}

export interface GetUserPostsRequest {
  handle?: Handle;
  pagination_key?: string;
  limit?: number;
}

export interface GetUserPostsResponse {
  posts: Post[];
  pagination_key: string;
}

export interface FollowUserRequest {
  handle: Handle;
}

export interface UnfollowUserRequest {
  handle: Handle;
}

export interface GetFollowStatusRequest {
  handle: Handle;
}

export interface FollowStatus {
  is_following: boolean;
  is_blocked: boolean;
  can_follow: boolean;
}

export interface GetMyHomeTimelineRequest {
  pagination_key?: string;
  limit?: number;
}

export interface MyHomeTimeline {
  posts: Post[];
  employer_posts: EmployerPost[];
  pagination_key: string;
}

export interface GetPostDetailsRequest {
  post_id: string;
}

export interface UpvoteUserPostRequest {
  post_id: string;
}

export interface DownvoteUserPostRequest {
  post_id: string;
}

export interface UnvoteUserPostRequest {
  post_id: string;
}

export interface FollowOrgRequest {
  domain: string;
}

export interface UnfollowOrgRequest {
  domain: string;
}

export interface GetEmployerPostDetailsRequest {
  employer_post_id: string;
}

File: typespec/hub/posts.go
package hub

import (
	"time"

	"github.com/vetchium/vetchium/typespec/common"
)

type AddPostRequest struct {
	Content string            `json:"content"  validate:"required,min=1,max=4096"`
	TagIDs  []common.VTagID   `json:"tag_ids"  validate:"max=3,dive,uuid"`
	NewTags []common.VTagName `json:"new_tags" validate:"max=3"`
}

type AddPostResponse struct {
	PostID string `json:"post_id"`
}

type Post struct {
	ID             string        `json:"id"`
	Content        string        `json:"content"`
	Tags           []string      `json:"tags"`
	AuthorName     string        `json:"author_name"`
	AuthorHandle   common.Handle `json:"author_handle"`
	CreatedAt      time.Time     `json:"created_at"`
	UpvotesCount   int32         `json:"upvotes_count"`
	DownvotesCount int32         `json:"downvotes_count"`
	Score          int32         `json:"score"`
	MeUpvoted      bool          `json:"me_upvoted"`
	MeDownvoted    bool          `json:"me_downvoted"`
	CanUpvote      bool          `json:"can_upvote"`
	CanDownvote    bool          `json:"can_downvote"`
	AmIAuthor      bool          `json:"am_i_author"`
}

type GetUserPostsRequest struct {
	Handle        *common.Handle `json:"handle"         validate:"omitempty,validate_handle"`
	PaginationKey *string        `json:"pagination_key" validate:"omitempty"`
	Limit         int            `json:"limit"          validate:"min=0,max=40"`
}

type GetUserPostsResponse struct {
	Posts         []Post `json:"posts"`
	PaginationKey string `json:"pagination_key"`
}

type FollowUserRequest struct {
	Handle common.Handle `json:"handle" validate:"required,validate_handle"`
}

type UnfollowUserRequest struct {
	Handle common.Handle `json:"handle" validate:"required,validate_handle"`
}

type GetFollowStatusRequest struct {
	Handle common.Handle `json:"handle" validate:"required,validate_handle"`
}

type FollowStatus struct {
	IsFollowing bool `json:"is_following"`
	IsBlocked   bool `json:"is_blocked"`
	CanFollow   bool `json:"can_follow"`
}

type GetMyHomeTimelineRequest struct {
	PaginationKey *string `json:"pagination_key" validate:"omitempty"`
	Limit         int     `json:"limit"          validate:"min=0,max=40"`
}

type MyHomeTimeline struct {
	Posts         []Post `json:"posts"`
	PaginationKey string `json:"pagination_key"`
}

type GetPostDetailsRequest struct {
	PostID string `json:"post_id" validate:"required"`
}

type UpvoteUserPostRequest struct {
	PostID string `json:"post_id" validate:"required"`
}

type DownvoteUserPostRequest struct {
	PostID string `json:"post_id" validate:"required"`
}

type UnvoteUserPostRequest struct {
	PostID string `json:"post_id" validate:"required"`
}
